### DISCUSSION TITLE
How do I extend the react aria props for auto-completion of attributes?

### DISCUSSION THREAD

**Discussion by PixeledCode** [adobe/react-spectrum#3631] [Q&A] [answered]
So, everything is working fine, except that typescript is not doing any auto-completion for any of the attributes that, for example, `useTextField` provides.

This is how I am using the hook:

```
import React from 'react';
import { useTextField } from 'react-aria';

function TextField(props) {
  let { label, isRequired } = props;
  let ref = React.useRef();
  let { labelProps, inputProps, descriptionProps, errorMessageProps } =
    useTextField(props, ref);

  return (
    <div>
      <label {...labelProps}>
        {label}
      </label>
      <input {...inputProps} ref={ref} required={props.isRequired || false} />
      {props.description && (
        <div {...descriptionProps}>{props.description}</div>
      )}
      {props.errorMessage && (
        <div {...errorMessageProps}>{props.errorMessage}</div>
      )}
    </div>
  );
}

export { TextField };
```

What changes do I need to make so that type is enabled for these props?

**Comment by reidbarber** [comment, marked as answer]
~~The link you posted isn't public. Would you be able to post a code snippet here?~~

**Update:**

You'll want to add a type to the props:

```jsx
import {AriaTextFieldProps} from '@react-types/textfield';

function TextField(props: AriaTextFieldProps) {
...
```

or extend them if you want to add your own custom props in addition to those:


```jsx
import {AriaTextFieldProps} from '@react-types/textfield';

interface TextFieldProps extends AriaTextFieldProps {
  // Your own custom props...
}

function TextField(props: TextFieldProps) {
...
```

You can also look at how we do this in React Spectrum components for inspiration.

**Reply by PixeledCode** [reply]
> The link you posted isn't public. Would you be able to post a code snippet here?

really sorry about that, I just edited the post to add the code.

**Reply by reidbarber** [reply]
no problem. Just updated my answer
